scipy/sparse/csgraph/_flow.pyx

```diff
@@ -670,3 +683,943 @@ cdef ITYPE_t[:] _dinic(
+        ITYPE_t[:] capacities,  # IN
+        ITYPE_t[:] demand,  # IN
+        ITYPE_t[:] cost,  # IN
+        ITYPE_t n_verts,  # IN
+        ITYPE_t[:] row,  # IN/OUT
+        ITYPE_t[:] col,  # IN/OUT
+        ITYPE_t[:] flow_data  # IN/OUT
+        ):
+    cdef:
+        ITYPE_t n_edges = capacities.shape[0]
+        ITYPE_t idx, faux_inf, capacities_sum, cost_sum
+        ITYPE_t j, s, t, i, tmp, p, q, n_non_zero_edges, v
+        ITYPE_t flow_cost, flow_value
+        ITYPE_t[:] edge_flow, vertex_potentials, Wne_len
+        ITYPE_t[:] parent, parent_edge, subtree_size
+        ITYPE_t[:] next_vertex_dft, prev_vertex_dft
+        ITYPE_t[:] last_descendent_dft
+        ITYPE_t[:] Wn, We, WnR, WeR, ancestors, subtree
+        return_struct r_struct
+        edge_result i_p_q
+        network_simplex_result ns_result = network_simplex_result()
+
+        ITYPE_t[:] edge_sources = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
+        ITYPE_t[:] edge_targets = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
+        ITYPE_t[:] edge_capacities = np.empty(n_edges + n_verts + 1,
+                                              dtype=ITYPE)
+        ITYPE_t[:] edge_weights = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
+
+    edge_flow = np.zeros(n_edges + n_verts, dtype=ITYPE)
+    vertex_potentials = np.empty(n_verts + 1, dtype=ITYPE)
+    parent = np.empty(n_verts + 1, dtype=ITYPE)
+    parent_edge = np.empty(n_edges + n_verts, dtype=ITYPE)
+    subtree_size = np.empty(n_verts + 1, dtype=ITYPE)
+    next_vertex_dft = np.empty(n_verts + 1, dtype=ITYPE)
+    prev_vertex_dft = np.empty(n_verts + 1, dtype=ITYPE)
+    last_descendent_dft = np.empty(n_verts + 1, dtype=ITYPE)
+    Wn = np.empty(n_edges + 1, dtype=ITYPE)
+    We = np.empty(n_edges + 1, dtype=ITYPE)
+    Wne_len = np.empty(2, dtype=ITYPE)
+    WnR = np.empty(n_verts, dtype=ITYPE)
+    WeR = np.empty(n_edges, dtype=ITYPE)
+    ancestors = np.empty(n_verts + 1, dtype=ITYPE)
+    subtree = np.empty(2*n_verts, dtype=ITYPE)
+    r_struct.prev_ret_value = False
+
+    with nogil:
+        idx = 0
+        n_non_zero_edges = 0
+        for e in range(n_edges):
+            if capacities[e] != 0:
+                n_non_zero_edges += 1
+                edge_sources[idx] = tails[e] + 1
+                edge_targets[idx] = heads[e] + 1
+                idx += 1
+        for v in range(n_verts):
+            if demand[v] > 0:
+                edge_sources[idx] = 0
+                edge_targets[idx] = v + 1
+            else:
+                edge_sources[idx] = v + 1
+                edge_targets[idx] = 0
+            idx += 1
+
+        capacities_sum, cost_sum = 0, 0
+        for e in range(n_edges):
+            if capacities[e] < INT_MAX:
+                capacities_sum += capacities[e]
+            cost_sum += abs(cost[e])
+        faux_inf = max(cost_sum, capacities_sum)
+        for v in range(n_verts):
+            faux_inf = max(faux_inf, demand[v])
+        faux_inf = 3*faux_inf or 1
```

Comment:
Why is `faux_inf` re-assigned this way?

Reply:
Taken from `networkx`,

https://github.com/networkx/networkx/blob/1620568e36702b1cfeaf1c0277b167b6cb93e48d/networkx/algorithms/flow/networksimplex.py#L559-L571